buildSrc/src/main/java/org/elasticsearch/gradle/info/GenerateGlobalBuildInfoTask.java

```diff
@@ -0,0 +1,270 @@
+package org.elasticsearch.gradle.info;
+
+import org.elasticsearch.gradle.OS;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.JavaVersion;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.tasks.CacheableTask;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputDirectory;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.internal.jvm.Jvm;
+import org.gradle.process.ExecResult;
+
+import javax.inject.Inject;
+import java.io.BufferedWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.List;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+@CacheableTask
+public class GenerateGlobalBuildInfoTask extends DefaultTask {
+    private JavaVersion minimumCompilerVersion;
+    private JavaVersion minimumRuntimeVersion;
+    private File compilerJavaHome;
+    private File runtimeJavaHome;
+    private List<JavaHome> javaVersions;
+    private final RegularFileProperty outputFile;
+    private final RegularFileProperty compilerVersionFile;
+    private final RegularFileProperty runtimeVersionFile;
+    private final RegularFileProperty fipsJvmFile;
+
+    @Inject
+    public GenerateGlobalBuildInfoTask(ObjectFactory objectFactory) {
+        this.outputFile = objectFactory.fileProperty();
+        this.compilerVersionFile = objectFactory.fileProperty();
+        this.runtimeVersionFile = objectFactory.fileProperty();
+        this.fipsJvmFile = objectFactory.fileProperty();
+    }
+
+    @TaskAction
+    public void generate() {
+        String javaVendor = System.getProperty("java.vendor");
+        String gradleJavaVersion = System.getProperty("java.version");
+        String gradleJavaVersionDetails = javaVendor + " " + gradleJavaVersion + " [" + System.getProperty("java.vm.name")
+            + " " + System.getProperty("java.vm.version") + "]";
+
+        String compilerJavaVersionDetails = gradleJavaVersionDetails;
+        JavaVersion compilerJavaVersionEnum = JavaVersion.current();
+        String runtimeJavaVersionDetails = gradleJavaVersionDetails;
+        JavaVersion runtimeJavaVersionEnum = JavaVersion.current();
+        File gradleJavaHome = Jvm.current().getJavaHome();
+        boolean inFipsJvm = false;
+
```

Comment:
We should do a check on compilerJavaHome and runtimeJava home somewhere in here to make sure they exist so we don't only run into an error running `jrunscript` this has generated a lot of confusion in the past